In [1]:
import os
import pandas as pd
import numpy as np

In [6]:
path = 'C:/Users/yuche/Dropbox/risk hedging/jupyter/original_data/year2018'

In [ ]:
path1 = 'C:/Users/yuche/Dropbox/risk hedging/jupyter/my_data'

In [7]:
os.chdir(path)

start processing DA price

In [8]:
#price on day-ahead market from SMART
price_file = pd.read_csv('day_ahead_price_2018.csv', sep = ";")
price_file.head()

Date Time of day Germany/Luxembourg[€/MWh]  Denmark 1[€/MWh]  \
0  31-Dec, 2017    12:00 AM                         -             26.98   
1  31-Dec, 2017     1:00 AM                         -             26.44   
2  31-Dec, 2017     2:00 AM                         -             26.20   
3  31-Dec, 2017     3:00 AM                         -             25.90   
4  31-Dec, 2017     4:00 AM                         -             25.37   

   Denmark 2[€/MWh]  France[€/MWh]  Northern Italy[€/MWh]  Netherlands[€/MWh]  \
0             26.98          10.94                  45.18               42.00   
1             26.44           9.04                  43.50               35.00   
2             26.20           6.88                  40.10               30.00   
3             25.90           3.69                  36.50               30.40   
4             25.37           3.26                  36.50               14.47   

  Poland[€/MWh]  Sweden 4[€/MWh]  Switzerland[€/MWh]  Slovenia[€/MWh]  \
0             -            26.98               38.94             4.90   
1             -            26.44               36.09             1.00   
2             -            26.20               31.99             0.02   
3             -            25.90               28.27            -3.71   
4             -            25.37               26.35             0.08   

   Czech Republic[€/MWh]  Hungary[€/MWh] Austria[€/MWh]  \
0                  -1.00            2.15              -   
1                  -0.10            0.21              -   
2                   0.64            0.64              -   
3                   2.15            2.15              -   
4                   3.06            3.06              -   

  Germany/Austria/Luxembourg[€/MWh],  
0                              -2.04  
1                              -4.99  
2                              -0.59  
3                              -3.71  
4                               0.08

In [9]:
price_file.tail()

Date Time of day Germany/Luxembourg[€/MWh]  Denmark 1[€/MWh]  \
8779  31-Dec, 2018     7:00 PM                     58.28             47.51   
8780  31-Dec, 2018     8:00 PM                     50.01             45.82   
8781  31-Dec, 2018     9:00 PM                     45.06             43.03   
8782  31-Dec, 2018    10:00 PM                     48.93             38.54   
8783  31-Dec, 2018    11:00 PM                     30.31             25.70   

      Denmark 2[€/MWh]  France[€/MWh]  Northern Italy[€/MWh]  \
8779             47.51          58.28                  70.06   
8780             45.82          50.01                  69.27   
8781             43.03          45.50                  59.84   
8782             38.54          49.80                  55.70   
8783             25.70          44.26                  49.80   

      Netherlands[€/MWh] Poland[€/MWh]  Sweden 4[€/MWh]  Switzerland[€/MWh]  \
8779               58.28             -            47.51               58.36   
8780               50.01             -            45.82               51.65   
8781               47.48             -            43.03               49.85   
8782               50.95             -            38.54               50.80   
8783               46.70             -            25.70               49.92   

      Slovenia[€/MWh]  Czech Republic[€/MWh]  Hungary[€/MWh] Austria[€/MWh]  \
8779            58.28                  55.34           55.34          58.28   
8780            50.01                  53.40           54.55          50.01   
8781            45.15                  49.77           50.95          45.15   
8782            55.01                  48.32           51.51          49.15   
8783            34.09                  43.00           45.00          34.09   

     Germany/Austria/Luxembourg[€/MWh],  
8779                                  -  
8780                                  -  
8781                                  -  
8782                                  -  
8783                                  -

In [6]:
test= price_file.iloc[24:,0:2]
test

Date Time of day
24     1-Jan, 2018    12:00 AM
25     1-Jan, 2018     1:00 AM
26     1-Jan, 2018     2:00 AM
27     1-Jan, 2018     3:00 AM
28     1-Jan, 2018     4:00 AM
...            ...         ...
8779  31-Dec, 2018     7:00 PM
8780  31-Dec, 2018     8:00 PM
8781  31-Dec, 2018     9:00 PM
8782  31-Dec, 2018    10:00 PM
8783  31-Dec, 2018    11:00 PM

[8760 rows x 2 columns]

In [7]:
price_file.loc[6574:6580]

Date Time of day Germany/Luxembourg[€/MWh]  Denmark 1[€/MWh]  \
6574  30-Sep, 2018    10:00 PM                         -             65.08   
6575  30-Sep, 2018    11:00 PM                         -             60.87   
6576   1-Oct, 2018    12:00 AM                     59.53             50.32   
6577   1-Oct, 2018     1:00 AM                      56.1             50.43   
6578   1-Oct, 2018     2:00 AM                     51.41             42.40   
6579   1-Oct, 2018     3:00 AM                     47.38             45.86   
6580   1-Oct, 2018     4:00 AM                     47.59             46.34   

      Denmark 2[€/MWh]  France[€/MWh]  Northern Italy[€/MWh]  \
6574             65.08          67.19                  72.96   
6575             60.87          61.19                  67.78   
6576             50.32          59.53                  63.11   
6577             50.43          56.10                  59.75   
6578             42.40          51.41                  59.35   
6579             45.86          47.38                  59.34   
6580             46.34          47.59                  59.38   

      Netherlands[€/MWh] Poland[€/MWh]  Sweden 4[€/MWh]  Switzerland[€/MWh]  \
6574               64.38             -            42.52               64.08   
6575               60.76             -            41.14               62.10   
6576               59.53             -            42.32               63.06   
6577               56.10             -            42.35               58.68   
6578               51.41             -            42.40               58.63   
6579               47.38             -            42.61               57.83   
6580               47.59             -            42.98               58.24   

      Slovenia[€/MWh]  Czech Republic[€/MWh]  Hungary[€/MWh] Austria[€/MWh]  \
6574            72.96                  58.63           62.08              -   
6575            60.87                  48.27           50.71              -   
6576            59.53                  43.92           43.92          59.53   
6577            56.10                  41.52           41.52           56.1   
6578            51.41                  37.50           37.50          51.41   
6579            47.38                  35.14           35.14          47.38   
6580            47.59                  39.81           39.81          47.59   

     Germany/Austria/Luxembourg[€/MWh],  
6574                              65.08  
6575                              60.87  
6576                                  -  
6577                                  -  
6578                                  -  
6579                                  -  
6580                                  -

In [8]:
array1 = np.array(price_file.loc[24:6575,'Germany/Austria/Luxembourg[€/MWh],'])
array1

array(['-5.27', '-29.99', '-56.65', ..., '65.31', '65.08', '60.87'],
      dtype=object)

In [9]:
array2 = np.array(price_file.loc[6576:8783,'Germany/Luxembourg[€/MWh]'])
array2

array(['59.53', '56.1', '51.41', ..., '45.06', '48.93', '30.31'],
      dtype=object)

In [10]:
print(array1.size, array2.size)

6552 2208


In [11]:
array = np.append(array1,array2)
array.size

8760

In [12]:
test["price"] = array
test

Date Time of day   price
24     1-Jan, 2018    12:00 AM   -5.27
25     1-Jan, 2018     1:00 AM  -29.99
26     1-Jan, 2018     2:00 AM  -56.65
27     1-Jan, 2018     3:00 AM  -63.14
28     1-Jan, 2018     4:00 AM  -64.62
...            ...         ...     ...
8779  31-Dec, 2018     7:00 PM   58.28
8780  31-Dec, 2018     8:00 PM   50.01
8781  31-Dec, 2018     9:00 PM   45.06
8782  31-Dec, 2018    10:00 PM   48.93
8783  31-Dec, 2018    11:00 PM   30.31

[8760 rows x 3 columns]

In [13]:
new = test[test.index % 24 == 0]
new["Time of day"].describe()

count          365
unique           1
top       12:00 AM
freq           365
Name: Time of day, dtype: object

In [14]:
test["price"].describe()

count     8760
unique    4473
top         50
freq        19
Name: price, dtype: object

In [15]:
test["price"] = test["price"].astype(float, errors = 'raise')
test["price"].describe()

count    8760.000000
mean       44.468492
std        17.771318
min       -76.010000
25%        34.455000
50%        45.090000
75%        54.870000
max       128.260000
Name: price, dtype: float64

In [16]:
test.reset_index(drop = True, inplace = True) 
test.rename(columns={"price": "day_price"}, inplace = True) 
test

Date Time of day  day_price
0      1-Jan, 2018    12:00 AM      -5.27
1      1-Jan, 2018     1:00 AM     -29.99
2      1-Jan, 2018     2:00 AM     -56.65
3      1-Jan, 2018     3:00 AM     -63.14
4      1-Jan, 2018     4:00 AM     -64.62
...            ...         ...        ...
8755  31-Dec, 2018     7:00 PM      58.28
8756  31-Dec, 2018     8:00 PM      50.01
8757  31-Dec, 2018     9:00 PM      45.06
8758  31-Dec, 2018    10:00 PM      48.93
8759  31-Dec, 2018    11:00 PM      30.31

[8760 rows x 3 columns]

#check yesterday price
test.loc[0:8759,:]

temp = test.iloc[24:8784,0:3]
temp.reset_index(drop = True, inplace = True)
temp

yes_price = test.iloc[0:8760,2].values
print(yes_price,len(yes_price))

temp['yes_price'] = yes_price
temp

temp.rename(columns={"price": "day_price"}, inplace = True) 
temp

start processing forecasted generation

In [17]:
#forecasted generation from ninja
fg_file = pd.read_csv('fg_pv_2018.csv')
fg_file.head(30)

local_time  electricity
0    1/1/2018 0:00        0.000
1    1/1/2018 1:00        0.000
2    1/1/2018 2:00        0.000
3    1/1/2018 3:00        0.000
4    1/1/2018 4:00        0.000
5    1/1/2018 5:00        0.000
6    1/1/2018 6:00        0.000
7    1/1/2018 7:00        0.000
8    1/1/2018 8:00        4.386
9    1/1/2018 9:00       43.322
10  1/1/2018 10:00       97.141
11  1/1/2018 11:00      137.736
12  1/1/2018 12:00      181.383
13  1/1/2018 13:00      265.931
14  1/1/2018 14:00      244.529
15  1/1/2018 15:00       88.206
16  1/1/2018 16:00        0.000
17  1/1/2018 17:00        0.000
18  1/1/2018 18:00        0.000
19  1/1/2018 19:00        0.000
20  1/1/2018 20:00        0.000
21  1/1/2018 21:00        0.000
22  1/1/2018 22:00        0.000
23  1/1/2018 23:00        0.000
24   1/2/2018 0:00        0.000
25   1/2/2018 1:00        0.000
26   1/2/2018 2:00        0.000
27   1/2/2018 3:00        0.000
28   1/2/2018 4:00        0.000
29   1/2/2018 5:00        0.000

In [18]:
fg_file['electricity'].describe()

count    8760.000000
mean      142.630198
std       215.526175
min         0.000000
25%         0.000000
50%         0.504500
75%       241.424000
max       809.182000
Name: electricity, dtype: float64

In [19]:
fg_file['electricity'] = fg_file['electricity'].apply(lambda x: x/1000).round(2)
fg_file.head(30)

local_time  electricity
0    1/1/2018 0:00         0.00
1    1/1/2018 1:00         0.00
2    1/1/2018 2:00         0.00
3    1/1/2018 3:00         0.00
4    1/1/2018 4:00         0.00
5    1/1/2018 5:00         0.00
6    1/1/2018 6:00         0.00
7    1/1/2018 7:00         0.00
8    1/1/2018 8:00         0.00
9    1/1/2018 9:00         0.04
10  1/1/2018 10:00         0.10
11  1/1/2018 11:00         0.14
12  1/1/2018 12:00         0.18
13  1/1/2018 13:00         0.27
14  1/1/2018 14:00         0.24
15  1/1/2018 15:00         0.09
16  1/1/2018 16:00         0.00
17  1/1/2018 17:00         0.00
18  1/1/2018 18:00         0.00
19  1/1/2018 19:00         0.00
20  1/1/2018 20:00         0.00
21  1/1/2018 21:00         0.00
22  1/1/2018 22:00         0.00
23  1/1/2018 23:00         0.00
24   1/2/2018 0:00         0.00
25   1/2/2018 1:00         0.00
26   1/2/2018 2:00         0.00
27   1/2/2018 3:00         0.00
28   1/2/2018 4:00         0.00
29   1/2/2018 5:00         0.00

In [20]:
fg = fg_file.iloc[0:8760,1].values
print(fg,len(fg))

[0. 0. 0. ... 0. 0. 0.] 8760


In [21]:
test['fore_gen'] = fg
test

Date Time of day  day_price  fore_gen
0      1-Jan, 2018    12:00 AM      -5.27       0.0
1      1-Jan, 2018     1:00 AM     -29.99       0.0
2      1-Jan, 2018     2:00 AM     -56.65       0.0
3      1-Jan, 2018     3:00 AM     -63.14       0.0
4      1-Jan, 2018     4:00 AM     -64.62       0.0
...            ...         ...        ...       ...
8755  31-Dec, 2018     7:00 PM      58.28       0.0
8756  31-Dec, 2018     8:00 PM      50.01       0.0
8757  31-Dec, 2018     9:00 PM      45.06       0.0
8758  31-Dec, 2018    10:00 PM      48.93       0.0
8759  31-Dec, 2018    11:00 PM      30.31       0.0

[8760 rows x 4 columns]

# adding forecasted price

In [22]:
test.shape[0]

8760

In [23]:
noise = np.random.normal(0, 5, test.shape[0])
noise

array([ -1.52911142,   0.03454076, -11.17108379, ...,   9.75918835,
        -4.9716349 ,   0.44679984])

In [24]:
test["fore_price"] = test["day_price"] + noise
test

Date Time of day  day_price  fore_gen  fore_price
0      1-Jan, 2018    12:00 AM      -5.27       0.0   -6.799111
1      1-Jan, 2018     1:00 AM     -29.99       0.0  -29.955459
2      1-Jan, 2018     2:00 AM     -56.65       0.0  -67.821084
3      1-Jan, 2018     3:00 AM     -63.14       0.0  -64.827428
4      1-Jan, 2018     4:00 AM     -64.62       0.0  -62.220562
...            ...         ...        ...       ...         ...
8755  31-Dec, 2018     7:00 PM      58.28       0.0   50.955990
8756  31-Dec, 2018     8:00 PM      50.01       0.0   58.396447
8757  31-Dec, 2018     9:00 PM      45.06       0.0   54.819188
8758  31-Dec, 2018    10:00 PM      48.93       0.0   43.958365
8759  31-Dec, 2018    11:00 PM      30.31       0.0   30.756800

[8760 rows x 5 columns]

# start converting hh, dd, mm

In [25]:
#convert 12 hour clock to 24 hour clock
def convert_hour(time):
    in_time = datetime.strptime(time, "%I:%M %p")
    out_time = datetime.strftime(in_time, "%H")
    out_num = int(out_time)
    return out_num
# convert date to month and day
def convert_month(date):
    in_time = datetime.strptime(date, "%d-%b, %Y")
    out_time = datetime.strftime(in_time, "%m")
    out_num = int(out_time)
    return out_num
def convert_day(date):
    in_time = datetime.strptime(date, "%d-%b, %Y")
    out_time = datetime.strftime(in_time, "%d")
    out_num = int(out_time)
    return out_num
def convert_week(date):
    in_time = datetime.strptime(date, "%d-%b, %Y")
    out_time = datetime.strftime(in_time, "%w")    
    return out_time

In [29]:
from datetime import datetime
test['hour'] = test['Time of day'].apply(lambda x: convert_hour(x))
test['month'] =test['Date'].apply(lambda x: convert_month(x))
test['day'] = test['Date'].apply(lambda x: convert_day(x))
test['weekday'] = test['Date'].apply(lambda x: convert_week(x))
test["week"] = test.index // 168 + 1
test.head(50)

Date Time of day  day_price  fore_gen  fore_price  hour  month  \
0   1-Jan, 2018    12:00 AM      -5.27      0.00   -6.799111     0      1   
1   1-Jan, 2018     1:00 AM     -29.99      0.00  -29.955459     1      1   
2   1-Jan, 2018     2:00 AM     -56.65      0.00  -67.821084     2      1   
3   1-Jan, 2018     3:00 AM     -63.14      0.00  -64.827428     3      1   
4   1-Jan, 2018     4:00 AM     -64.62      0.00  -62.220562     4      1   
5   1-Jan, 2018     5:00 AM     -67.00      0.00  -77.340037     5      1   
6   1-Jan, 2018     6:00 AM     -72.54      0.00  -64.224983     6      1   
7   1-Jan, 2018     7:00 AM     -76.01      0.00  -90.222193     7      1   
8   1-Jan, 2018     8:00 AM     -71.45      0.00  -66.173712     8      1   
9   1-Jan, 2018     9:00 AM     -66.88      0.04  -64.025304     9      1   
10  1-Jan, 2018    10:00 AM     -62.00      0.10  -69.388356    10      1   
11  1-Jan, 2018    11:00 AM     -56.08      0.14  -61.858816    11      1   
12  1-Jan, 2018    12:00 PM     -49.96      0.18  -58.299275    12      1   
13  1-Jan, 2018     1:00 PM     -24.29      0.27  -27.094583    13      1   
14  1-Jan, 2018     2:00 PM      -5.59      0.24    4.365633    14      1   
15  1-Jan, 2018     3:00 PM       0.23      0.09   -6.833593    15      1   
16  1-Jan, 2018     4:00 PM      11.02      0.00   12.130230    16      1   
17  1-Jan, 2018     5:00 PM      23.50      0.00   24.432217    17      1   
18  1-Jan, 2018     6:00 PM      22.91      0.00   30.487059    18      1   
19  1-Jan, 2018     7:00 PM      21.02      0.00   18.486820    19      1   
20  1-Jan, 2018     8:00 PM      21.05      0.00   24.645503    20      1   
21  1-Jan, 2018     9:00 PM      21.95      0.00   10.726360    21      1   
22  1-Jan, 2018    10:00 PM      23.52      0.00   24.128835    22      1   
23  1-Jan, 2018    11:00 PM      18.96      0.00   20.883430    23      1   
24  2-Jan, 2018    12:00 AM      18.12      0.00   26.629892     0      1   
25  2-Jan, 2018     1:00 AM      14.99      0.00   21.445133     1      1   
26  2-Jan, 2018     2:00 AM      13.63      0.00   14.742202     2      1   
27  2-Jan, 2018     3:00 AM      12.15      0.00   20.438330     3      1   
28  2-Jan, 2018     4:00 AM      11.83      0.00   12.962900     4      1   
29  2-Jan, 2018     5:00 AM      14.55      0.00   19.537178     5      1   
30  2-Jan, 2018     6:00 AM      22.58      0.00   27.429996     6      1   
31  2-Jan, 2018     7:00 AM      41.92      0.00   49.486488     7      1   
32  2-Jan, 2018     8:00 AM      47.95      0.00   40.744887     8      1   
33  2-Jan, 2018     9:00 AM      49.90      0.03   45.295897     9      1   
34  2-Jan, 2018    10:00 AM      50.00      0.05   46.339409    10      1   
35  2-Jan, 2018    11:00 AM      49.70      0.06   50.198627    11      1   
36  2-Jan, 2018    12:00 PM      48.16      0.07   43.806757    12      1   
37  2-Jan, 2018     1:00 PM      47.24      0.07   42.381840    13      1   
38  2-Jan, 2018     2:00 PM      46.10      0.04   50.900554    14      1   
39  2-Jan, 2018     3:00 PM      43.86      0.03   49.186458    15      1   
40  2-Jan, 2018     4:00 PM      43.64      0.00   44.132899    16      1   
41  2-Jan, 2018     5:00 PM      47.37      0.00   30.673356    17      1   
42  2-Jan, 2018     6:00 PM      46.08      0.00   42.694169    18      1   
43  2-Jan, 2018     7:00 PM      41.08      0.00   34.328890    19      1   
44  2-Jan, 2018     8:00 PM      34.94      0.00   32.069620    20      1   
45  2-Jan, 2018     9:00 PM      25.46      0.00   35.478062    21      1   
46  2-Jan, 2018    10:00 PM      21.30      0.00   16.180163    22      1   
47  2-Jan, 2018    11:00 PM       9.94      0.00    6.335274    23      1   
48  3-Jan, 2018    12:00 AM       7.93      0.00   -3.837351     0      1   
49  3-Jan, 2018     1:00 AM       0.03      0.00    4.746063     1      1   

    day weekday  week  
0     1       1     1  
1     1       1     1  
2     1    

In [30]:
test.tail(50)

Date Time of day  day_price  fore_gen  fore_price  hour  month  \
8710  29-Dec, 2018    10:00 PM      24.84      0.00   28.064990    22     12   
8711  29-Dec, 2018    11:00 PM      -3.71      0.00  -13.512009    23     12   
8712  30-Dec, 2018    12:00 AM      13.75      0.00   14.067190     0     12   
8713  30-Dec, 2018     1:00 AM       0.02      0.00   -0.163111     1     12   
8714  30-Dec, 2018     2:00 AM       5.88      0.00    7.630948     2     12   
8715  30-Dec, 2018     3:00 AM       2.09      0.00   12.369378     3     12   
8716  30-Dec, 2018     4:00 AM       1.36      0.00    7.012009     4     12   
8717  30-Dec, 2018     5:00 AM      -9.90      0.00   -6.297535     5     12   
8718  30-Dec, 2018     6:00 AM     -16.77      0.00  -20.299342     6     12   
8719  30-Dec, 2018     7:00 AM     -19.43      0.00  -20.683534     7     12   
8720  30-Dec, 2018     8:00 AM      30.94      0.00   23.309037     8     12   
8721  30-Dec, 2018     9:00 AM      42.10      0.03   42.644317     9     12   
8722  30-Dec, 2018    10:00 AM      49.54      0.08   48.319176    10     12   
8723  30-Dec, 2018    11:00 AM      51.16      0.14   48.557549    11     12   
8724  30-Dec, 2018    12:00 PM      51.16      0.20   46.550050    12     12   
8725  30-Dec, 2018     1:00 PM      49.64      0.18   39.515446    13     12   
8726  30-Dec, 2018     2:00 PM      46.19      0.07   44.722073    14     12   
8727  30-Dec, 2018     3:00 PM      47.64      0.01   46.881199    15     12   
8728  30-Dec, 2018     4:00 PM      48.33      0.00   51.662477    16     12   
8729  30-Dec, 2018     5:00 PM      58.40      0.00   65.957586    17     12   
8730  30-Dec, 2018     6:00 PM      62.43      0.00   62.399056    18     12   
8731  30-Dec, 2018     7:00 PM      62.94      0.00   59.906595    19     12   
8732  30-Dec, 2018     8:00 PM      62.42      0.00   53.206507    20     12   
8733  30-Dec, 2018     9:00 PM      54.28      0.00   49.211130    21     12   
8734  30-Dec, 2018    10:00 PM      55.63      0.00   59.525309    22     12   
8735  30-Dec, 2018    11:00 PM      55.00      0.00   54.053625    23     12   
8736  31-Dec, 2018    12:00 AM      50.94      0.00   58.944908     0     12   
8737  31-Dec, 2018     1:00 AM      49.57      0.00   52.150319     1     12   
8738  31-Dec, 2018     2:00 AM      48.32      0.00   47.738294     2     12   
8739  31-Dec, 2018     3:00 AM      45.89      0.00   54.861839     3     12   
8740  31-Dec, 2018     4:00 AM      45.47      0.00   46.569996     4     12   
8741  31-Dec, 2018     5:00 AM      47.11      0.00   45.243432     5     12   
8742  31-Dec, 2018     6:00 AM      49.12      0.00   49.449943     6     12   
8743  31-Dec, 2018     7:00 AM      53.79      0.00   54.747935     7     12   
8744  31-Dec, 2018     8:00 AM      57.47      0.00   52.924980     8     12   
8745  31-Dec, 2018     9:00 AM      61.52      0.02   58.521245     9     12   
8746  31-Dec, 2018    10:00 AM      62.64      0.06   63.385363    10     12   
8747  31-Dec, 2018    11:00 AM      64.56      0.07   54.838363    11     12   
8748  31-Dec, 2018    12:00 PM      65.01      0.07   74.746805    12     12   
8749  31-Dec, 2018     1:00 PM      63.91      0.05   56.416882    13     12   
8750  31-Dec, 2018     2:00 PM      59.68      0.02   59.262678    14     12   
8751  31-Dec, 2018     3:00 PM      56.60      0.00   54.100050    15     12   
8752  31-Dec, 2018     4:00 PM      63.28      0.00   68.503816    16     12   
8753  31-Dec, 2018     5:00 PM      68.01      0.00   60.646015    17     12   
8754  31-Dec, 2018     6:00 PM      66.98      0.00   72.189120    18     12   
8755  31-Dec, 2018     7:00 PM      58.28      0.00   50.955990    19     12   
8756  31-Dec, 2018     8:00 PM      50.01      0.00   58.396447    20     12   
8757  31-Dec, 2018     9:00 PM      45.06      0.00   54.819188    21     12   
8758  31-Dec, 2018    10:00 PM      48.93      0.00   43.958365    22     12   
8759  31-Dec, 

In [32]:
test.rename(columns={"day_price": "act_price"}, inplace = True)
test

Date Time of day  act_price  fore_gen  fore_price  hour  month  \
0      1-Jan, 2018    12:00 AM      -5.27       0.0   -6.799111     0      1   
1      1-Jan, 2018     1:00 AM     -29.99       0.0  -29.955459     1      1   
2      1-Jan, 2018     2:00 AM     -56.65       0.0  -67.821084     2      1   
3      1-Jan, 2018     3:00 AM     -63.14       0.0  -64.827428     3      1   
4      1-Jan, 2018     4:00 AM     -64.62       0.0  -62.220562     4      1   
...            ...         ...        ...       ...         ...   ...    ...   
8755  31-Dec, 2018     7:00 PM      58.28       0.0   50.955990    19     12   
8756  31-Dec, 2018     8:00 PM      50.01       0.0   58.396447    20     12   
8757  31-Dec, 2018     9:00 PM      45.06       0.0   54.819188    21     12   
8758  31-Dec, 2018    10:00 PM      48.93       0.0   43.958365    22     12   
8759  31-Dec, 2018    11:00 PM      30.31       0.0   30.756800    23     12   

      day weekday  week  
0       1       1     1  
1       1       1     1  
2       1       1     1  
3       1       1     1  
4       1       1     1  
...   ...     ...   ...  
8755   31       1    53  
8756   31       1    53  
8757   31       1    53  
8758   31       1    53  
8759   31       1    53  

[8760 rows x 10 columns]

In [34]:
test.to_csv("da_2018_un.csv", index = None)

# starting normalization of the data 

In [39]:
data = test.iloc[:,2:-1]
data

act_price  fore_gen  fore_price  hour  month  day weekday
0         -5.27       0.0   -6.799111     0      1    1       1
1        -29.99       0.0  -29.955459     1      1    1       1
2        -56.65       0.0  -67.821084     2      1    1       1
3        -63.14       0.0  -64.827428     3      1    1       1
4        -64.62       0.0  -62.220562     4      1    1       1
...         ...       ...         ...   ...    ...  ...     ...
8755      58.28       0.0   50.955990    19     12   31       1
8756      50.01       0.0   58.396447    20     12   31       1
8757      45.06       0.0   54.819188    21     12   31       1
8758      48.93       0.0   43.958365    22     12   31       1
8759      30.31       0.0   30.756800    23     12   31       1

[8760 rows x 7 columns]

In [40]:
from sklearn import preprocessing
scaler = preprocessing.MinMaxScaler()
names = data.columns
d = scaler.fit_transform(data)
scaled_data = pd.DataFrame(d, columns=names)
scaled_data.head()

act_price  fore_gen  fore_price      hour  month  day   weekday
0   0.346306       0.0    0.375715  0.000000    0.0  0.0  0.166667
1   0.225290       0.0    0.271425  0.043478    0.0  0.0  0.166667
2   0.094777       0.0    0.100888  0.086957    0.0  0.0  0.166667
3   0.063005       0.0    0.114371  0.130435    0.0  0.0  0.166667
4   0.055760       0.0    0.126112  0.173913    0.0  0.0  0.166667

In [41]:
scaled_data.describe()

act_price     fore_gen   fore_price         hour        month  \
count  8760.000000  8760.000000  8760.000000  8760.000000  8760.000000   
mean      0.589800     0.176074     0.606754     0.500000     0.502366   
std       0.086999     0.266157     0.083228     0.300982     0.313459   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.540779     0.000000     0.559789     0.250000     0.272727   
50%       0.592843     0.000000     0.609579     0.500000     0.545455   
75%       0.640721     0.296296     0.656686     0.750000     0.818182   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               day      weekday  
count  8760.000000  8760.000000  
mean      0.490685     0.499087  
std       0.293225     0.333351  
min       0.000000     0.000000  
25%       0.233333     0.166667  
50%       0.500000     0.500000  
75%       0.733333     0.833333  
max       1.000000     1.000000

In [42]:
scaled_data.rename(columns={
                     "act_price": "norm_ap",
                     "fore_gen": "norm_fg",
                     "fore_price": "norm_fp",
                     "hour": "norm_hh",
                     "month": "norm_mm",
                     "day": "norm_dd",
                     "weekday": "norm_ww"}, inplace = True)
scaled_data

norm_ap  norm_fg   norm_fp   norm_hh  norm_mm  norm_dd   norm_ww
0     0.346306      0.0  0.375715  0.000000      0.0      0.0  0.166667
1     0.225290      0.0  0.271425  0.043478      0.0      0.0  0.166667
2     0.094777      0.0  0.100888  0.086957      0.0      0.0  0.166667
3     0.063005      0.0  0.114371  0.130435      0.0      0.0  0.166667
4     0.055760      0.0  0.126112  0.173913      0.0      0.0  0.166667
...        ...      ...       ...       ...      ...      ...       ...
8755  0.657414      0.0  0.635828  0.826087      1.0      1.0  0.166667
8756  0.616929      0.0  0.669337  0.869565      1.0      1.0  0.166667
8757  0.592696      0.0  0.653226  0.913043      1.0      1.0  0.166667
8758  0.611641      0.0  0.604312  0.956522      1.0      1.0  0.166667
8759  0.520488      0.0  0.544856  1.000000      1.0      1.0  0.166667

[8760 rows x 7 columns]

In [43]:
norm = pd.concat([test, scaled_data], axis=1)
norm

Date Time of day  act_price  fore_gen  fore_price  hour  month  \
0      1-Jan, 2018    12:00 AM      -5.27       0.0   -6.799111     0      1   
1      1-Jan, 2018     1:00 AM     -29.99       0.0  -29.955459     1      1   
2      1-Jan, 2018     2:00 AM     -56.65       0.0  -67.821084     2      1   
3      1-Jan, 2018     3:00 AM     -63.14       0.0  -64.827428     3      1   
4      1-Jan, 2018     4:00 AM     -64.62       0.0  -62.220562     4      1   
...            ...         ...        ...       ...         ...   ...    ...   
8755  31-Dec, 2018     7:00 PM      58.28       0.0   50.955990    19     12   
8756  31-Dec, 2018     8:00 PM      50.01       0.0   58.396447    20     12   
8757  31-Dec, 2018     9:00 PM      45.06       0.0   54.819188    21     12   
8758  31-Dec, 2018    10:00 PM      48.93       0.0   43.958365    22     12   
8759  31-Dec, 2018    11:00 PM      30.31       0.0   30.756800    23     12   

      day weekday  week   norm_ap  norm_fg   norm_fp   norm_hh  norm_mm  \
0       1       1     1  0.346306      0.0  0.375715  0.000000      0.0   
1       1       1     1  0.225290      0.0  0.271425  0.043478      0.0   
2       1       1     1  0.094777      0.0  0.100888  0.086957      0.0   
3       1       1     1  0.063005      0.0  0.114371  0.130435      0.0   
4       1       1     1  0.055760      0.0  0.126112  0.173913      0.0   
...   ...     ...   ...       ...      ...       ...       ...      ...   
8755   31       1    53  0.657414      0.0  0.635828  0.826087      1.0   
8756   31       1    53  0.616929      0.0  0.669337  0.869565      1.0   
8757   31       1    53  0.592696      0.0  0.653226  0.913043      1.0   
8758   31       1    53  0.611641      0.0  0.604312  0.956522      1.0   
8759   31       1    53  0.520488      0.0  0.544856  1.000000      1.0   

      norm_dd   norm_ww  
0         0.0  0.166667  
1         0.0  0.166667  
2         0.0  0.166667  
3         0.0  0.166667  
4         0.0  0.166667  
...       ...       ...  
8755      1.0  0.166667  
8756      1.0  0.166667  
8757      1.0  0.166667  
8758      1.0  0.166667  
8759      1.0  0.166667  

[8760 rows x 17 columns]

In [44]:
df = norm.loc[norm.index % 24 == 0]
df["Time of day"].describe()

count          365
unique           1
top       12:00 AM
freq           365
Name: Time of day, dtype: object

In [45]:
norm.to_csv("da_2018_norm.csv", index = None)